File: internal/discord/session.go
package discord

import (
	"strings"

	"github.com/bwmarrin/discordgo"

	"github.com/cscercel/doty-lore-bot/internal/db"
	"github.com/cscercel/doty-lore-bot/internal/handler"
)

func RegisterHandlers(s *discordgo.Session, q *db.Queries) {
	s.AddHandler(func(s *discordgo.Session, i *discordgo.InteractionCreate) {
		switch i.Type {
		case discordgo.InteractionApplicationCommand:
			data := i.ApplicationCommandData()
			if data.Name != "lore" || len(data.Options) == 0 {
				return
			}
			sub := data.Options[0]
			switch sub.Name {
			case "create":
				handler.HandleCreateStart(s, i, sub.Options, q)
			case "view":
				handler.HandleView(s, i, sub.Options, q)
			case "list":
				handler.HandleList(s, i, sub.Options, q)
			case "edit":
				handler.HandleEditStart(s, i, sub.Options, q)
			case "delete":
				handler.HandleDelete(s, i, sub.Options, q)
			}

		case discordgo.InteractionApplicationCommandAutocomplete:
			data := i.ApplicationCommandData()
			if data.Name != "lore" || len(data.Options) == 0 {
				return
			}
			sub := data.Options[0]
			if sub.Name == "view" || sub.Name == "edit" || sub.Name == "delete" {
				handler.HandleViewAutocomplete(s, i, sub.Options, q)
			}

		case discordgo.InteractionModalSubmit:
			customID := i.ModalSubmitData().CustomID
			switch {
			case strings.HasPrefix(customID, "lore_create_modal:"):
				handler.HandleCreateSubmit(s, i, q)
			case strings.HasPrefix(customID, "lore_edit_modal:"):
				handler.HandleEditSubmit(s, i, q)
			}
		}
	})
}

File: src/session.ts
import { Client, Events, Interaction } from "discord.js";
import { handleCreateStart, handleCreateSubmit } from "./handlers/create";
import { handleDelete } from "./handlers/delete";
import { handleEditStart, handleEditSubmit } from "./handlers/edit";
import { handleList } from "./handlers/list";
import { handleViewAutoComplete, handleView } from "./handlers/view";


export function registerHandlers(client: Client) {
    client.on(Events.InteractionCreate, async (interaction: Interaction) => {
        // Slash commands
        if (interaction.isChatInputCommand()) {
            if (interaction.commandName !== "lore") return;

            const sub = interaction.options.getSubcommand(false);
            if (!sub) return;

            switch (sub) {
                case "create":
                    await handleCreateStart(interaction);
                    break;
                case "view":
                    await handleView(interaction);
                    break;
                case "list":
                    await handleList(interaction);
                    break;
                case "edit":
                    await handleEditStart(interaction);
                    break;
                case "delete":
                    await handleDelete(interaction);
                    break;
            }
            return;
        }

        // Autocomplete
        if (interaction.isAutocomplete()) {
            if (interaction.commandName !== "lore") return;

            const sub = interaction.options.getSubcommand(false);
            if (!sub) return;

            if (sub === "view" || sub === "edit" || sub === "delete") {
                await handleViewAutoComplete(interaction);
            }
            return;
        }

        // Modal submissions
        if (interaction.isModalSubmit()) {
            const customId = interaction.customId;

            if (customId.startsWith("lore_create_modal:")) {
                await handleCreateSubmit(interaction);
            } else if (customId.startsWith("lore_edit_modal:")) {
                await handleEditSubmit(interaction);
            }
            return;
        }
    });
}
